### PR TITLE
Native TensorFlow input queue implemented with fit, predict, evaluate (TFRecords)

### PR DESCRIPTION
**Pull request by colinskow**
Adds support for native multi-threaded QueueRunner / TFRecord input using the TensorFlow backend through the `fit`, `predict`, and `evaluate` functions. This includes data validation.

Working demo: https://github.com/colinskow/keras/blob/fit-tensor-input/examples/mnist_tfrecord.py

Each batch of input is evaluated inside the training loop and fed into the trainer. There are ways to enhance the performance by feeding input directly into the graph, but this is going to mean a LOT of work to decrease training time perhaps a few percent. Let's get the functionality implemented and work on increasing performance under the hood down the road.

WHY THIS IS NEEDED:

Processing input via Python/NumPy is very slow. Especially when paying for a high-end GPU, inputting through Python can become a bottleneck which takes longer than actual training. Keras' input pipeline is not cloud friendly.

TensorFlow has a very good solution to these problems, but it does not work with Keras now.
https://www.tensorflow.org/programmers_guide/reading_data

This PR allows Keras to train, evaluate, and predict using TF's super-fast, cloud friendly, multi-threaded input pipeline. This is essential to running models in the cloud and not wasting money when paying by the minute for high-end GPUs. And it does this transparently with no breaking changes to the API. (And other backends can be accommodated in the same way if they have similar functionality.)

HELP REQUESTED:
1) I'm relatively new to Python, so please tell me if I've used any bad coding practices
2) TODO: automated testing of TFRecord input, including possible edge cases

@fchollet @ahundt @TimZaman

### CODE REVIEW

File: keras/models.py [hunk 1]
```diff
@@ -790,40 +790,78 @@ def compile(self, optimizer, loss,
...
-                1 for progress bar logging, 2 for one log line per epoch.
-            callbacks: list of `keras.callbacks.Callback` instances.
-                List of callbacks to apply during training.
+                Not required when using tensors as input.
+            steps_per_epoch: Only relevant when using tensors as input.
+                Total number of steps (batches of samples)
+                to evaluate from input tensors before declaring one epoch
+                finished and starting the next epoch. It should typically
+                be equal to the number of unique samples if your dataset
+                divided by the batch size.
```
Comment (TimZaman) [inline keras/models.py]: This falls flat when the division does not result in a discrete integer.

File: keras/models.py [hunk 2]
```diff
@@ -790,40 +790,78 @@ def compile(self, optimizer, loss,
         self.sample_weights = self.model.sample_weights
         self.targets = self.model.targets
 
-    def fit(self, x, y, batch_size=32, epochs=10, verbose=1, callbacks=None,
-            validation_split=0., validation_data=None, shuffle=True,
-            class_weight=None, sample_weight=None, initial_epoch=0, **kwargs):
+    def fit(self, x=None,
```
Comment (TimZaman) [inline keras/models.py]: Why the style changes?

File: examples/mnist_tfrecord.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+def count_tfrecord_entries(filename):
+    c = 0
+    for record in tf.python_io.tf_record_iterator(filename):
+        c += 1
+    return c
+
+
+def cnn_model(input_shape, num_classes):
+    model = Sequential()
```
Comment (TimZaman) [inline examples/mnist_tfrecord.py]: Try using the functional model API.

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -395,6 +413,18 @@ def is_keras_tensor(x):
...
+def is_native_tensor(x):
+    """Returns whether `x` is a native tensor.
+
+        # Arguments
+            x: a potential tensor.
+
+        # Returns
+            A boolean: whether the argument is a native tensor.
+    """
+    return isinstance(x, tf.Tensor)
```
Comment (TimZaman) [inline keras/backend/tensorflow_backend.py]: What about `tf.SparseTensor`, `tf.Variable`?
Anyway, you can just use the existing `is_keras_tensor()`.

File: keras/backend/cntk_backend.py [hunk 5]
```diff
@@ -263,6 +263,19 @@ def is_keras_tensor(x):
     return hasattr(x, '_keras_history')
 
 
+def is_native_tensor(x):
```
Comment (TimZaman) [inline keras/backend/cntk_backend.py]: use `is_keras_tensor()`

File: keras/engine/training.py [hunk 6]
```diff
@@ -1119,46 +1194,71 @@ def _predict_loop(self, f, ins, batch_size=32, verbose=0):
...
             Scalar loss (if the model has a single output and no metrics)
             or list of scalars (if the model has multiple outputs
             and/or metrics). The attribute `model.metrics_names` will give you
             the display labels for the scalar outputs.
         """
-        if ins and hasattr(ins[0], 'shape'):
+        tensor_input = _is_tf_tensor(ins[0])
+        if ins and tensor_input:
+            if not num_steps:
+                raise ValueError('`validation_steps` must be specified when using tensor input.')
```
Comment (colinskow) [inline keras/engine/training.py]: Change `validation_steps` to `num_steps`

File: keras/backend/theano_backend.py [hunk 7]
```diff
@@ -201,6 +201,18 @@ def is_keras_tensor(x):
     return hasattr(x, '_keras_history')
 
 
+def is_native_tensor(x):
```
Comment (TimZaman) [inline keras/backend/theano_backend.py]: use `is_keras_tensor()`

File: keras/engine/training.py [hunk 8]
```diff
@@ -964,9 +978,38 @@ def _make_predict_function(self):
                                                name='predict_function',
                                                **kwargs)
```
Comment (colinskow) [inline keras/engine/training.py]: Since these functions below are platform specific, perhaps they should be implemented in the backend instead?

File: keras/models.py (hunk 2 above)
Comment (colinskow) [inline keras/models.py]: I copied and pasted from the fit function of `training.py`. After adding parameters the long lines were awkward.

File: keras/models.py (hunk 1 above)
Comment (colinskow) [inline keras/models.py]: This is copied directly from the `fit_generator` docs.

File: keras/backend/tensorflow_backend.py (hunk 4 above)
Comment (colinskow) [inline keras/backend/tensorflow_backend.py]: `is_keras_tensor()` doesn't work. This does. `tf.train.shuffle_batch` returns type `tf.Tensor`.

File: keras/backend/theano_backend.py (hunk 7 above)
Comment (colinskow) [inline keras/backend/theano_backend.py]: DOESN'T WORK. There is no Keras metadata in the input tensor.

File: examples/mnist_tfrecord.py (hunk 3 above)
Comment (colinskow) [inline examples/mnist_tfrecord.py]: Copied straight from `examples/mnist_cnn.py`. I appreciate constructive feedback, but feel you are criticizing to criticize.

File: examples/mnist_tfrecord.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                'depth': _int64_feature(depth),
+                'label': _int64_feature(int(labels[index])),
+                'image_raw': _bytes_feature(image_raw)}))
+            writer.write(example.SerializeToString())
+        writer.close()
+    else:
+        print('tfrecord %s already exists' % filename)
+
+
+def read_and_decode_recordinput(tf_glob, one_hot=True, classes=None, is_train=None, batch_shape=[128, 28, 28, 1]):
```
Comment (colinskow) [inline examples/mnist_tfrecord.py]: This function can be simplified quite a bit based on this example:
https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/examples/how_tos/reading_data/fully_connected_reader.py#L48

Comment (colinskow) [inline examples/mnist_tfrecord.py]: @ahundt is the expert in this area.

File: keras/backend/theano_backend.py (hunk 7 above)
Comment (Dref360) [inline keras/backend/theano_backend.py]: InputLayer will have some metadatas once https://github.com/fchollet/keras/pull/7066 is merged.

File: keras/backend/tensorflow_backend.py (hunk 4 above)
Comment (Dref360) [inline keras/backend/tensorflow_backend.py]: Yeah but your function will get used in the future by others and they expect is to work for all types of Tensors.

**Comment by Dref360** [comment]
This PR is really TF specific and adds a lot of code into the main files only for TF (training.py for example). Not convinced this is needed.

**Comment by colinskow** [comment]
@Dref360 see the discussion toward the bottom of #6928.

TL;DR processing input via Python/NumPy is very slow. Especially when paying for a high-end GPU, inputting through Python can become a bottleneck which takes longer than actual training. Keras' input pipeline is not cloud friendly.

TensorFlow has a very good solution to these problems, but it does not work with Keras now.
https://www.tensorflow.org/programmers_guide/reading_data

This PR allows Keras to train, evaluate, and predict using TF's super-fast, cloud friendly, multi-threaded input pipeline. This is essential to running models in the cloud and not wasting money when paying by the minute for high-end GPUs. And it does this transparently with no breaking changes to the API.

Unfortunately I haven't seen that Theano or CNTK support similar functionality. If they do we can definitely accommodate them all with the same API.

I do agree `training.py` is an ugly place for TF-specific code. I'm open to feedback on how to best refactor this into the backend. But this is very important functionality for running models in production where performance is critical.

File: keras/backend/tensorflow_backend.py (hunk 4 above)
Comment (colinskow) [inline keras/backend/tensorflow_backend.py]: Understood, I will copy the functionality from `_is_keras_tensor()` but remove the check for Keras metadata.

File: keras/models.py (hunk 1 above)
Comment (TimZaman) [inline keras/models.py]: Yet the tfrecords pipe you have in your example does not support a non-fixed batch size, while you are not sure your dataset size is not cleanly divisible by the batch size. Having a super fast data pipe in tf and having mixed-size batch sizes is hard. Optimized Tensorflow models therefore often works with steps, rather than epochs.

File: examples/mnist_tfrecord.py (hunk 3 above)
Comment (TimZaman) [inline examples/mnist_tfrecord.py]: Nonsense. Just trying to help.

**Comment by TimZaman** [comment]
Agree with @Dref360 .

> TL;DR processing input via Python/NumPy is very slow.

Yet you are making a `tf->np->tf` roundtrip, where `->np->` is in the critical path.
For feeding the GPU, it's all about the critical path. Keras's python-multiprocessing pipeline feeds fine and can keep up with things in principle. The problem  you'll encounter on training on big systems like a [DGX-1](https://www.nvidia.com/en-us/data-center/dgx-1/), is that the GPUs are extremely fast. The GPU's don't want to wait for anything to get the next batch, and the samples should be staged on the GPU. They should not be staged on CPU, and they should not be waiting on a `tf->np->tf` roundtrip to finish.

The first step in doing this is having a tensorflow input, this is exactly what #6928 was in the process of solving. This not dependent on TFRecords, yield ops, etc.
The feature we want is: _"Having Keras cleanly support input $framework tensors instead of numpy arrays"_.

**Comment by colinskow** [comment]
@TimZaman I agree with you that #6928 is the most performance-effecient solution. The problem is that it requires refactoring the user's model and doesn't work with most features of the fit API such as data validation and sample weights. There's a lot of work and problem solving ahead to make it merge ready.

My solution is ready to go (except for some fine-tuning and cleanup) and works out of the box with all existing models. The only API change is specifying the number of steps per epoch (as with `fit_generator`). It gives 95% of the performance benefits of a native input queue. And it will be easy to implement with other backends when they develop input queue solutions as well.

I think you are exaggerating the performance impact of tf->np->tf once per batch. The training loop is doing this anyway. Just look how much Python code already gets executed inside `_fit_loop` between batches. The GPU is waiting anyway. The TF QueueRunner has the input processed and ready to go in memory when Python calls for it. The cost of feeding a NP.array back into TF is miniscule. And this bottleneck can be compensated for by simply increasing batch size.

To really get the performance you are talking about means eliminating Python from the training loop completely and going all-native. And the performance benefit will likely be only a few percent. I don't think this is going to be implemented anytime soon.

I think it's a good idea to actually benchmark my solution against #6928 to see what the difference really is. I personally only have a CPU, so perhaps someone with a good GPU can try too.

I do suggest we merge this first (after cleaning it up, adding tests etc.) Then when the technical kinks are worked out of #6928 we can merge that down the road to really get the extra performance.

And I don't agree with depriving users of a very useful feature for months while waiting on a slightly better implementation when we can release now and then transparently tune under the hood.

**Comment by colinskow** [comment]
I benchmarked this pull request against @ahundt's solution in #6829 on my 2012 MacBook Pro (CPU only). Here are the results:

This uses the same data, same model architecture, and same TFRecords input pipeline. I used a batch size of 200 across 60,000 samples for 5 epochs. I disabled data validation to be fair since #6829 doesn't support it.

```
[ahundt/keras@tfrecords] (counts data samples)
Epoch 1/5
60000/60000 [==============================] - 120s - loss: 0.2943 - acc: 0.9110      
Epoch 2/5
60000/60000 [==============================] - 111s - loss: 0.0868 - acc: 0.9742     
Epoch 3/5
60000/60000 [==============================] - 111s - loss: 0.0648 - acc: 0.9807     
Epoch 4/5
60000/60000 [==============================] - 111s - loss: 0.0556 - acc: 0.9839     
Epoch 5/5
60000/60000 [==============================] - 111s - loss: 0.0490 - acc: 0.9850


[colinskow/keras@fit-tensor-input] (counts batches)
Epoch 1/5
300/300 [==============================] - 115s - loss: 0.2585 - acc: 0.9215     
Epoch 2/5
300/300 [==============================] - 114s - loss: 0.0869 - acc: 0.9745     
Epoch 3/5
300/300 [==============================] - 114s - loss: 0.0639 - acc: 0.9815     
Epoch 4/5
300/300 [==============================] - 113s - loss: 0.0515 - acc: 0.9848     
Epoch 5/5
300/300 [==============================] - 114s - loss: 0.0452 - acc: 0.9867
```

So 111 seconds vs 114 seconds. That's a difference of about 2.7%. Going from TF->NP-> TF costs about 10 ms / batch on my MacBook. And for me it is a cost worth paying to have full use of the `fit`, `predict`, and `validate` APIs.

My proposal is that we go with this API since it is much simpler, and down the road figure out how to plug the inputs directly into the graph under the hood to get the 2.7% performance boost.

**Comment by TimZaman** [comment]
Who uses CPU for DL? This MR won't benefit anyone with 1 CPU or 1 or 2 GPUs. The tensor-native input mostly makes sense for GPU heavy systems, as Keras's multiprocessing is already quite good to feed 1 or 2 GPUs. I am repeating myself so I will refrain from further comment.

**Comment by colinskow** [comment]
@TimZaman I agree with you 100% on that native tensor input is ideal. All I'm suggesting is that we get the functionality implemented first and then make it performant later since that is the hard part.

File: keras/engine/training.py [hunk 10]
```diff
@@ -26,6 +26,18 @@
 from ..legacy import interfaces
 
 
+def _is_tf_tensor(x):
```
Comment (ahundt) [inline keras/engine/training.py]: This and everything that uses it is too tf specific for this file. See how I handled these same issues in #6928

**Comment by fchollet** [comment]
At some point, we want to be able to call models' `fit` on data tensors. 

But there are only two reasonable ways to make it happen:
- 1) Graph rewiring: replacing the graph's placeholders with the data tensors for the duration of `fit` (then restore them back)
- 2) Rebuilding the same graph on top of the data tensors

The present proposal, at a high level, extracts batches from the data tensors (one memory copy) then feeds them to the placeholders. It's not as crazy as it sounds, but still very inefficient.

Importantly, this approach can be achieved with nearly no changes to the codebase (compared to the fairly involved changes proposed here) by simply structuring them as a Python generator that would be used with `fit_generator`. The generator would run the data tensors in a different session, and yield the resulting Numpy arrays -- that's it. This would be self-contained and would only involve a few lines of code.

Overall, we should not proceed with this approach and we should keep exploring 1) and 2).

**Comment by ahundt** [comment]
> ii. Rebuilding the same graph on top of the data tensors

This is the current approach after the latest commits in #6928 (with a more atomic merge process outlined in https://github.com/fchollet/keras/pull/7072#issuecomment-311174255).

**Comment by colinskow** [comment]
Thanks for the feedback @fchollet I agree with your points. Extracting the feed data inside a `fit_generator` is a great idea, I didn't think about that. Although the ideal solution is definitely wiring the tensors directly into the graph.

For my personal projects I'm looking into plugging directly into the TensorFlow `estimator` and `experiment` APIs since these are built for easily taking models into distributed production.

I'll go ahead and close this so I'm not cluttering up your pending PRs.

**Comment by ahundt** [comment]
@colinskow While you're exploring could you consider creating an mnist example (or any other kind) with keras models, an estimator & experiment and submit a PR here or directly in TF? I think that would help many users!
